core/src/main/java/org/elasticsearch/search/internal/InternalSearchHit.java

```diff
@@ -514,6 +544,185 @@ public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) t
         return builder;
     }
 
+    public static InternalSearchHit fromXContent(XContentParser parser) throws IOException {
+        XContentParser.Token token;
+        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser::getTokenLocation);
+        String currentFieldName = null;
+        String type = null, id = null;
+        Text index = null;
+        float score = DEFAULT_SCORE;
+        long version = -1;
+        SearchSortValues sortValues = SearchSortValues.EMPTY;
+        InternalNestedIdentity nestedIdentity = null;
+        Map<String, HighlightField> highlightFields = new HashMap<>();
+        BytesReference parsedSource = null;
+        List<String> matchedQueries = new ArrayList<>();
+        Map<String, SearchHitField> fields = new HashMap<>();
+        Explanation explanation = null;
+        ShardId shardId = null;
+        String nodeId = null;
+        Map<String, InternalSearchHits> innerHits = null;
+        while((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+                if (Fields.SORT.equals(currentFieldName)) {
+                    sortValues = SearchSortValues.fromXContent(parser);
+                } else if (InternalNestedIdentity.Fields._NESTED.equals(currentFieldName)) {
+                    nestedIdentity = InternalNestedIdentity.fromXContent(parser);
+                }
+            } else if (token.isValue()) {
+                if (Fields._TYPE.equals(currentFieldName)) {
+                    type = parser.text();
+                } else if (Fields._INDEX.equals(currentFieldName)) {
+                    index = new Text(parser.text());
+                } else if (Fields._ID.equals(currentFieldName)) {
+                    id = parser.text();
+                } else if (Fields._SCORE.equals(currentFieldName)) {
+                     score = parser.floatValue();
+                } else if (Fields._VERSION.equals(currentFieldName)) {
+                    version = parser.longValue();
+                } else if (Fields._SHARD.equals(currentFieldName)) {
+                    String shard = parser.text();
+                    shardId = parseShardId(shard, parser.getTokenLocation());
+                } else if (Fields._NODE.equals(currentFieldName)) {
+                   nodeId = parser.text();
+                } else if (MapperService.isMetadataField(currentFieldName)) {
+                    List<Object> values = new ArrayList<>();
+                    values.add(parseValue(parser));
```

Comment:
In this case 'list()' wouldn't work I think, because the meta data fields (like '_parent' etc.) are not written as an array but as a single value. Another idea would be to access AbstractXContentParser#readValue() for this, but I remember some related discussion about controlling which token types are parser how that we had for GetField, which is where I took the parseValue() method from. Maybe @javanna has an opinion here as well.